merkle/src/lib.rs

```diff
@@ -46,21 +74,68 @@ impl<F: Field, H: Hasher<F>> MerkleTree<F, H> {
     /// # Returns
     ///
     /// Returns an error type for easier "monadic style" chaining.
-    #[allow(dead_code)]
-    #[allow(clippy::result_unit_err)]
     pub fn verify_decommitment(
         &self,
         comm: H::Output,
         to_verify: &[(usize, H::Output)],
-        siblings: impl Iterator<Item = H::Output>,
-    ) -> Result<(), ()> {
-        if self.root_decommitment(to_verify, siblings).ok_or(())? == comm {
-            Ok(())
+    ) -> Result<bool, MerkleError> {
+        assert!(!to_verify.is_empty(), "Empty verify queries");
+
+        let queries_idx: Vec<usize> = to_verify.iter().map(|&(x, _)| x).collect();
+        let siblings = self
+            .generate_decommitment(&queries_idx)
+            .ok_or(MerkleError::InvalidQuery)?;
+        if self
+            .root_decommitment(to_verify, siblings.into_iter())
+            .ok_or(MerkleError::VerificationFail)?
+            == comm
+        {
+            Ok(true)
         } else {
-            Err(())
+            Ok(false)
         }
     }
 
+    /// Generates siblings from input query indices.
+    ///
+    /// # Arguments
+    ///
+    /// - `queries_idx`: an array of [position] to query
+    ///
+    /// # Returns
+    ///
+    /// - `Some(sib)`: vector of sibling hashes
+    fn generate_decommitment(&self, queries_idx: &[usize]) -> Option<Vec<H::Output>> {
```

Comment:
I think it might be useful to separate the Merkle tree verifier logic, from the Merkle tree prover logic: 

The verifier will basically only have the root and the number of expected levels. His only operation is verifying (batch) openings.

The prover will have all the nodes in the tree and be able to generate decommitments.

Reply:
The logic of Verifier and Prover is now separate, which is using the channel implementation.